### PR TITLE
Add Deepseek MoE v1 & GigaChat models

### PR DESCRIPTION
**Pull request by Inf1delis**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)

Self-reported review complexity:
- [x]  Medium

The PR adds support for DeepSeek MoE v1 models ([Base](https://huggingface.co/deepseek-ai/deepseek-moe-16b-base) and [Instruct](https://huggingface.co/deepseek-ai/deepseek-moe-16b-chat)) & support new GigaChat models ([Base](https://huggingface.co/ai-sage/GigaChat-20B-A3B-base) and [Instruct](https://huggingface.co/ai-sage/GigaChat-20B-A3B-instruct)). Since GigaChat is based on the Deepseek MoE v1 architecture, the changes for that model is limited to the tokenizer.

### CODE REVIEW

**Review by alterfo — APPROVED** [review]

**Comment by Inf1delis** [comment]
@ggerganov Hi! I think this PR is ready, could you check it up?

File: convert_hf_to_gguf.py [hunk 1]
```diff
@@ -3506,6 +3509,97 @@ def prepare_tensors(self):
                 raise ValueError(f"Unprocessed experts: {experts}")
 
 
+@Model.register("DeepseekForCausalLM")
+class DeepseekModel(Model):
```
Comment (ggerganov) [inline convert_hf_to_gguf.py:3513]: Move before `DeepseekV2Model` above

File: src/llama.cpp [hunk 2]
```diff
@@ -8907,6 +8959,55 @@ static bool llm_load_tensors(
...
+                            layer.ffn_up_exps   = create_tensor(tn(LLM_TENSOR_FFN_UP_EXPS,   "weight", i), {  n_embd, n_ff_exp, n_expert}, 0);
+
+                            // Shared expert branch
+                            layer.ffn_gate_shexp = create_tensor(tn(LLM_TENSOR_FFN_GATE_SHEXP, "weight", i), {n_embd, n_ff_exp * n_expert_shared}, 0);
+                            layer.ffn_down_shexp = create_tensor(tn(LLM_TENSOR_FFN_DOWN_SHEXP, "weight", i), {        n_ff_exp * n_expert_shared, n_embd}, 0);
+                            layer.ffn_up_shexp   = create_tensor(tn(LLM_TENSOR_FFN_UP_SHEXP,   "weight", i), {n_embd, n_ff_exp * n_expert_shared}, 0);
+                        }
+                    }
+                } break;
+            case LLM_ARCH_DEEPSEEK:
```
Comment (ggerganov) [inline src/llama.cpp:8988]: Move before `case LLM_ARCH_DEEPSEEK2` above.

**Review by ggerganov — APPROVED** [review]
Minor suggestions to fix the location of the new DS code to be located before DS2

**Comment by Inf1delis** [comment]
Thank you for your suggestions! I hadn't noticed that.  
The changes have been made: the new DS code is now placed before `DeepseekV2Model` and before case `LLM_ARCH_DEEPSEEK2`.

**Pull request merged**